extension/awsmiddleware/middleware_test.go

```diff
@@ -0,0 +1,169 @@
+	handleResponse func(r *http.Response)
+	start          time.Time
+	end            time.Time
+}
+
+var _ RequestHandler = (*testHandler)(nil)
+var _ ResponseHandler = (*testHandler)(nil)
+
+func (t *testHandler) ID() string {
+	return t.id
+}
+
+func (t *testHandler) Position() HandlerPosition {
+	return t.position
+}
+
+func (t *testHandler) HandleRequest(r *http.Request) {
+	t.start = time.Now()
+	if t.handleRequest != nil {
+		t.handleRequest(r)
+	}
+}
+
+func (t *testHandler) HandleResponse(r *http.Response) {
+	t.end = time.Now()
+	if t.handleResponse != nil {
+		t.handleResponse(r)
+	}
+}
+
+func (t *testHandler) Latency() time.Duration {
+	return t.end.Sub(t.start)
+}
+
+func TestHandlerPosition(t *testing.T) {
+	testCases := []struct {
+		position HandlerPosition
+		str      string
+	}{
+		{position: After, str: "after"},
+		{position: Before, str: "before"},
+	}
+	for _, testCase := range testCases {
+		position := testCase.position
+		got, err := position.MarshalText()
+		assert.NoError(t, err)
+		assert.EqualValues(t, testCase.str, got)
+		assert.NoError(t, position.UnmarshalText(got))
+		assert.Equal(t, position, testCase.position)
+	}
+}
+
+func TestInvalidHandlerPosition(t *testing.T) {
+	position := HandlerPosition(-1)
+	got, err := position.MarshalText()
+	assert.Error(t, err)
+	assert.ErrorIs(t, err, errUnsupportedPosition)
+	assert.Nil(t, got)
+	err = position.UnmarshalText([]byte("HandlerPosition(-1)"))
+	assert.Error(t, err)
+	assert.ErrorIs(t, err, errUnsupportedPosition)
+}
+
+func TestInvalidHandlers(t *testing.T) {
+	invalidHandler := &testHandler{id: "invalid handler", position: -1}
+	testExtension := &testMiddlewareExtension{
+		requestHandlers:  []RequestHandler{invalidHandler},
+		responseHandlers: []ResponseHandler{invalidHandler},
+	}
+	// v1
+	client := awstesting.NewClient()
+	err := ConfigureSDKv1(testExtension, &client.Handlers)
+	assert.Error(t, err)
+	assert.True(t, errors.Is(err, errInvalidHandler))
+	assert.True(t, errors.Is(err, errUnsupportedPosition))
+	// v2
+	err = ConfigureSDKv2(testExtension, &awsv2.Config{})
+	assert.Error(t, err)
+	assert.True(t, errors.Is(err, errInvalidHandler))
+	assert.True(t, errors.Is(err, errUnsupportedPosition))
+}
+
+func TestConfigureSDKv1(t *testing.T) {
+	middleware, recorder, server := setup(t)
+	client := awstesting.NewClient(&awsv1.Config{
+		Region:     awsv1.String("mock-region"),
+		DisableSSL: awsv1.Bool(true),
+		Endpoint:   awsv1.String(server.URL),
+	})
+	require.Equal(t, 3, client.Handlers.Build.Len())
+	require.Equal(t, 0, client.Handlers.Unmarshal.Len())
+	assert.NoError(t, ConfigureSDKv1(middleware, &client.Handlers))
+	assert.Equal(t, 5, client.Handlers.Build.Len())
+	assert.Equal(t, 1, client.Handlers.Unmarshal.Len())
+	s3Client := &s3v1.S3{Client: client}
+	output, err := s3Client.ListBuckets(&s3v1.ListBucketsInput{})
+	require.NoError(t, err)
+	assert.NotNil(t, output)
+	assert.GreaterOrEqual(t, recorder.Latency(), testLatency)
```

Comment:
It's checking that the end time (when the handler got the response) occurred at least after the expected latency. I don't think it can be flaky because the test server always sleeps for that amount of time, which will prevent the response from happening before then.

Reply:
Ah I see. I missed the sleep.